File: frontend/src/services/coupon-service.ts
import api from "@/lib/api";
import type { ApiResponse } from "@/types/api";

export interface Coupon {
  id: number;
  code: string;
  type: string;
  value: number;
  minOrderAmount: number;
  maxDiscount: number | null;
  discountAmount: number;
  valid: boolean;
  message: string;
}

/**
 * Validate a coupon code.
 */
export async function validateCoupon(
  code: string,
  orderAmount: number
): Promise<ApiResponse<Coupon>> {
  const response = await api.get(`/coupons/${code}`, {
    params: { orderAmount },
  });
  return response.data;
}

/**
 * Apply a coupon (increment usage count).
 */
export async function applyCoupon(code: string): Promise<ApiResponse<void>> {
  const response = await api.post(`/coupons/${code}/apply`);
  return response.data;
}

File: backend/src/main/java/com/headless/ecommerce/service/CouponService.java
package com.headless.ecommerce.service;

import com.headless.ecommerce.dto.response.CouponResponse;

import java.math.BigDecimal;

/**
 * Service interface for coupon management.
 */
public interface CouponService {

    /**
     * Validate and get coupon details.
     */
    CouponResponse validateCoupon(String code, BigDecimal orderAmount);

    /**
     * Apply a coupon (increment usage count).
     */
    void applyCoupon(String code);
}
